apps/meteor/server/models/raw/Users.js

```diff
@@ -270,18 +262,115 @@ export class UsersRaw extends BaseRaw {
 		return this.findPaginated(query, options);
 	}
 
+	findPaginatedActiveUsersByRoomIdWithHighestRole(
+		searchTerm,
+		rid,
+		searchFields,
+		options = {},
+		extraQuery = [],
+		{ startsWith = false, endsWith = false } = {},
+	) {
+		const regexString = (startsWith ? '^' : '') + escapeRegExp(searchTerm) + (endsWith ? '$' : '');
+		const termRegex = new RegExp(regexString, 'i');
+
+		const orStmt = (searchFields || []).map((el) => ({ [el.trim()]: termRegex }));
+
+		const query = {
+			$and: [
+				{
+					active: true,
+					__rooms: rid,
+					username: { $exists: true },
+					// if the search term is empty, don't need to have the $or statement (because it would be an empty regex)
+					...(searchTerm && orStmt.length && { $or: orStmt }),
+				},
+				...extraQuery,
+			],
+		};
+
+		const skip =
+			options.skip !== 0
+				? [
+						{
+							$skip: options.skip,
+						},
+				  ]
+				: [];
+
+		const limit =
+			options.limit !== 0
+				? [
+						{
+							$limit: options.limit,
+						},
+				  ]
+				: [];
+
+		return this.col.aggregate([
+			{
+				$match: query,
+			},
+			{
+				$lookup: {
+					from: 'rocketchat_subscription',
+					localField: '_id',
+					foreignField: 'u._id',
+					as: 'sub',
+					pipeline: [
+						{
+							$match: { $expr: { $eq: ['$rid', rid] } },
+						},
+					],
+				},
+			},
+			{ $unwind: '$sub' },
+			{
+				$project: {
+					...options.projection,
+					roles: '$sub.roles',
+				},
+			},
+			{
+				$addFields: {
+					highestRole: {
+						$cond: [
+							{ $in: ['owner', '$roles'] },
+							{ role: 'owner', level: 0 },
+							{ $cond: [{ $in: ['moderator', '$roles'] }, { role: 'moderator', level: 1 }, { role: 'member', level: 2 }] },
+						],
+					},
+				},
+			},
+			{
+				$facet: {
+					members: [
+						{
+							$sort: {
+								'highestRole.level': 1,
+								...options.sort,
+							},
+						},
+						...skip,
+						...limit,
+					],
+					totalCount: [{ $group: { _id: null, total: { $sum: 1 } } }],
+				},
+			},
+		]);
```

Comment:
This query could potentially operate on a ton of results, it may need `allowDiskUse` option to be passed